src/rollup/rollup.ts

```
@@ -79,11 +79,21 @@ export async function rollupInternal(
 			if (watchFiles.length > 0) {
 				error_.watchFiles = watchFiles;
 			}
-			await graph.pluginDriver.hookParallel('buildEnd', [error_]);
-			await graph.pluginDriver.hookParallel('closeBundle', []);
+			try {
+				await graph.pluginDriver.hookParallel('buildEnd', [error_]);
+			} catch (buildEndError: any) {
+				await graph.pluginDriver.hookParallel('closeBundle', [buildEndError]);
```

I wonder if we should pass the `buildEndError`, the original error, or maybe something different here? Also, one issue someone had that I vaguely remember is that the buildEndError could hide the original error, which makes debugging harder. On the other hand, we should not omit it, as the logic in buildEnd is obviously faulty.
One nice solution could be to throw a "concatenated error", e.g.

"There was an error during the build: <message>. Additionally, handling the error in buildEnd caused another error: <message>"

And set the cause of the error to the first error. What do you think?

---

I just tried this, and indeed the original error gets erased in the current version of rollup:

```js
const plugin = {
  name: "example",
  resolveId(id) {
    this.error('Original error');
  },
  buildEnd(error) {
    // If error is not handled here, it gets erased
    this.error('Another error');
  },
};
```

Maybe this could be solved by:

- Automatically setting `.cause` on the error (if not defined), so that the chain is preserved. This might be wrong because we artificially create a chain of errors without knowing the plugin author's intention.
- Instead of receiving the last error, `closeBundle` receives an array of all accumulated errors. This would partly solve the issue you mentioned:

```js
const plugin = {
  name: "example",
  resolveId(id) {
    this.error('Original error');
  },
  buildEnd(error) {
    // If error is not handled here, it gets erased
    this.error('Another error');
  },
  bundleClose(errors) {
    if (errors.length === 0) return;
    console.error('Build failed with', errors.length, 'errors');
    for (const error of errors) console.error(error);
  }
};
```

I'm not very familiar with rollup architecture regarding inter-plugin communication; would would all plugins receive all errors? Would they only receive their errors?

In the meantime, we could update the documentation to include the fact that `buildEnd` has the ability to erase previous errors, and that `cause` should be used to preserve them.

---

Here is a draft of said documentation update: https://github.com/rollup/rollup/pull/5868

---

What about my suggestion of throwing a new Error that contains both error messages and has one of the errors as "cause"? Then we would still lose the stack trace of the other error, but we would at least learn that both errors occurred? Having an array would be problematic because it might require bigger changes in other parts to properly display this, e.g. in Rollup CLI. Then this feature would also become a bugfix of sorts...

---

At second glance it seems like a sensible middle-ground, little to implem and addresses all painpoints. I'll update this PR to see if the pieces fall together!

---

I managed to make something that works! If the wording is ok to you, this is ready to merge 👌